File: java-algorithm/src/test/java/com/example/algorithm/graph/DijkstraTest.java
package com.example.algorithm.graph;

import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.junit.jupiter.MockitoExtension;

import static org.junit.jupiter.api.Assertions.assertArrayEquals;
import static org.junit.jupiter.api.Assertions.assertEquals;

@ExtendWith(MockitoExtension.class)
@Slf4j
class DijkstraTest {
    Dijkstra.Graph graph = new Dijkstra.Graph();
    Dijkstra.Vertex vertexS;
    Dijkstra.Vertex vertexT;
    Dijkstra.Vertex vertexX;
    Dijkstra.Vertex vertexY;
    Dijkstra.Vertex vertexZ;

    @BeforeEach
    void setUp() {
        vertexS = new Dijkstra.Vertex("S");
        vertexT = new Dijkstra.Vertex("T");
        vertexX = new Dijkstra.Vertex("X");
        vertexY = new Dijkstra.Vertex("Y");
        vertexZ = new Dijkstra.Vertex("Z");
        graph.addVertex(vertexS);
        graph.addVertex(vertexT);
        graph.addVertex(vertexX);
        graph.addVertex(vertexY);
        graph.addVertex(vertexZ);
        graph.addEdge(vertexS, vertexT, 10);
        graph.addEdge(vertexS, vertexY, 5);
        graph.addEdge(vertexT, vertexX, 1);
        graph.addEdge(vertexT, vertexY, 2);
        graph.addEdge(vertexX, vertexZ, 4);
        graph.addEdge(vertexY, vertexT, 3);
        graph.addEdge(vertexY, vertexX, 9);
        graph.addEdge(vertexY, vertexZ, 2);
        graph.addEdge(vertexZ, vertexS, 7);
        graph.addEdge(vertexZ, vertexX, 6);
    }

    @AfterEach
    void tearDown() {
    }

    @Test
    void testDijkstraAlgorithm() {
        Dijkstra.dijkstraAlgorithm(graph, vertexS);
        var pathT = Dijkstra.getShortestPath(vertexT);
        var pathX = Dijkstra.getShortestPath(vertexX);
        var pathY = Dijkstra.getShortestPath(vertexY);
        var pathZ = Dijkstra.getShortestPath(vertexZ);
        assertEquals(3, pathT.length);
        assertEquals(4, pathX.length);
        assertEquals(2, pathY.length);
        assertEquals(3, pathZ.length);
        assertArrayEquals(new Dijkstra.Vertex[]{vertexS, vertexY, vertexT}, pathT);
        assertArrayEquals(new Dijkstra.Vertex[]{vertexS, vertexY, vertexT, vertexX}, pathX);
        assertArrayEquals(new Dijkstra.Vertex[]{vertexS, vertexY}, pathY);
        assertArrayEquals(new Dijkstra.Vertex[]{vertexS, vertexY, vertexZ}, pathZ);
        assertEquals(8, Dijkstra.getShortestDistance(vertexT));
        assertEquals(9, Dijkstra.getShortestDistance(vertexX));
        assertEquals(5, Dijkstra.getShortestDistance(vertexY));
        assertEquals(7, Dijkstra.getShortestDistance(vertexZ));
    }
}

File: cpp-algorithm/src/graph/test/dijkstra_test.cpp
#include "dijkstra.h"

#include <gtest/gtest.h>

GTEST_TEST(Dijkstra, Simple1)
{
    auto vertex_s = Dijkstra::Vertex('S');
    auto vertex_t = Dijkstra::Vertex('T');
    auto vertex_x = Dijkstra::Vertex('X');
    auto vertex_y = Dijkstra::Vertex('Y');
    auto vertex_z = Dijkstra::Vertex('Z');

    const auto graph = new Dijkstra::Graph();

    graph->AddVertex(vertex_s);
    graph->AddVertex(vertex_t);
    graph->AddVertex(vertex_x);
    graph->AddVertex(vertex_y);
    graph->AddVertex(vertex_z);

    // S
    graph->AddEdge(vertex_s, vertex_t, 10);
    graph->AddEdge(vertex_s, vertex_y, 5);
    // T
    graph->AddEdge(vertex_t, vertex_x, 1);
    graph->AddEdge(vertex_t, vertex_y, 2);
    // X
    graph->AddEdge(vertex_x, vertex_z, 4);
    // Y
    graph->AddEdge(vertex_y, vertex_t, 3);
    graph->AddEdge(vertex_y, vertex_x, 9);
    graph->AddEdge(vertex_y, vertex_z, 2);
    // Z
    graph->AddEdge(vertex_z, vertex_s, 7);
    graph->AddEdge(vertex_z, vertex_x, 6);

    auto& source = vertex_s;
    source.distance = 0;
    graph->DijkstraAlgorithm(source);

    // weight
    ASSERT_EQ(0, vertex_s.distance);
    ASSERT_EQ(8, vertex_t.distance);
    ASSERT_EQ(9, vertex_x.distance);
    ASSERT_EQ(5, vertex_y.distance);
    ASSERT_EQ(7, vertex_z.distance);

    // predecessor
    ASSERT_EQ('Y', vertex_t.predecessor->id);
    ASSERT_EQ('T', vertex_x.predecessor->id);
    ASSERT_EQ('S', vertex_y.predecessor->id);
    ASSERT_EQ('Y', vertex_z.predecessor->id);
}
